deducers.go

```diff
@@ -0,0 +1,292 @@
+	select {
+	case sa := <-retchans.ret:
+		return sa, nil
+	case err := <-retchans.err:
+		return nil, err
+	}
+}
+
+type sourceCoordinator struct {
+	callMgr   *callManager
+	srcmut    sync.RWMutex // guards srcs and nameToURL maps
+	srcs      map[string]*sourceGateway
+	nameToURL map[string]string
+	psrcmut   sync.Mutex // guards protoSrcs map
+	protoSrcs map[string][]srcReturnChans
+	deducer   *deductionCoordinator
+	cachedir  string
+}
+
+func newSourceCoordinator(cm *callManager, deducer *deductionCoordinator, cachedir string) *sourceCoordinator {
+	return &sourceCoordinator{
+		callMgr:   cm,
+		deducer:   deducer,
+		cachedir:  cachedir,
+		srcs:      make(map[string]*sourceGateway),
+		nameToURL: make(map[string]string),
+		protoSrcs: make(map[string][]srcReturnChans),
+	}
+}
+
+func (sc *sourceCoordinator) getSourceGatewayFor(ctx context.Context, id ProjectIdentifier) (*sourceGateway, error) {
+	normalizedName := id.normalizedSource()
+
+	sc.srcmut.RLock()
+	if url, has := sc.nameToURL[normalizedName]; has {
+		if srcGate, has := sc.srcs[url]; has {
+			sc.srcmut.RUnlock()
+			return srcGate, nil
+		}
+	}
+	sc.srcmut.RUnlock()
+
+	// No gateway exists for this path yet; set up a proto, being careful to fold
+	// together simultaneous attempts on the same path.
+	rc := srcReturnChans{
+		ret: make(chan *sourceGateway),
+		err: make(chan error),
+	}
+
+	// The rest of the work needs its own goroutine, the results of which will
+	// be re-joined to this call via the return chans.
+	go func() {
+		sc.psrcmut.Lock()
+		if chans, has := sc.protoSrcs[normalizedName]; has {
+			// Another goroutine is already working on this normalizedName. Fold
+			// in with that work by attaching our return channels to the list.
+			sc.protoSrcs[normalizedName] = append(chans, rc)
+			sc.psrcmut.Unlock()
+			return
+		}
+
+		sc.protoSrcs[normalizedName] = []srcReturnChans{rc}
+		sc.psrcmut.Unlock()
+
+		doReturn := func(sa *sourceGateway, err error) {
+			sc.psrcmut.Lock()
+			if sa != nil {
+				for _, rc := range sc.protoSrcs[normalizedName] {
+					rc.ret <- sa
+				}
+			} else if err != nil {
+				for _, rc := range sc.protoSrcs[normalizedName] {
+					rc.err <- err
+				}
+			} else {
+				panic("sa and err both nil")
+			}
+
+			delete(sc.protoSrcs, normalizedName)
+			sc.psrcmut.Unlock()
+		}
+
+		pd, err := sc.deducer.deduceRootPath(normalizedName)
+		if err != nil {
+			// As in the deducer, don't cache errors so that externally-driven retry
+			// strategies can be constructed.
+			doReturn(nil, err)
+			return
+		}
+
+		// It'd be quite the feat - but not impossible - for an actor
+		// corresponding to this normalizedName to have slid into the main
+		// sources map after the initial unlock, but before this goroutine got
+		// scheduled. Guard against that by checking the main sources map again
+		// and bailing out if we find an entry.
+		sc.srcmut.RLock()
+		srcGate, has := sc.srcs[normalizedName]
+		sc.srcmut.RUnlock()
+		if has {
```

Comment:
No argument here! I'd definitely like to see this decomposed into more comprehensible pieces. IMO it's already less complected than it was in the old system (where root deduction and source creation were inextricably bound up in those stacks of futures), but there's still a ways to go.